File: src/main/java/game/Maze.java
package game;

import misc.Constants;
import misc.PointNode;

import java.awt.Point;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * The class that generates mazes.
 */
public class Maze {
    /**
     * Constructs a new Maze that generates a game
     * with the given number of rows and columns.
     */
    public Maze() {
        tiles_ = new boolean[Constants.Game.ROWS][Constants.Game.COLUMNS];
        init();
    }

    /**
     * Checks if a certain point is an open pathway in the maze.
     *
     * @param p the point
     * @return idk fam what do you think
     */
    public boolean hasPathAt(Point p) {
        return contains(p) && tiles()[p.y][p.x];
    }

    /**
     * Checks if a point lies within the generated mazes.
     *
     * @param p the point
     * @return whether or not a point lies in this maze
     */
    public static boolean contains(Point p) {
        return (p.x | p.y) >= 0
            && p.x < Constants.Game.COLUMNS
            && p.y < Constants.Game.ROWS;
    }

    public boolean[][] tiles() {
        return tiles_;
    }

    private boolean[][] tiles_;

    private void init() {
        generatePaths();
        patchDeadEnds();
    }

    private void generatePaths() {
        List<Point> stack = new ArrayList<>();
        Point first = new Point(0, 0);
        stack.add(first);
        tiles_[first.y][first.x] = true;

        while (!stack.isEmpty()) {
            Point current = stack.get(stack.size() - 1);
            Point[] nexts = nextPaths(current);
            if (nexts.length > 0) {
                Point next = nexts[(int) (Math.random() * nexts.length)];
                makePath(current, next);
                stack.add(next);
            } else {
                stack.remove(stack.size() - 1);
            }
        }
    }

    private void patchDeadEnds() {
        for (int y = 0; y < tiles().length; y += 2) {
            for (int x = 0; x < tiles()[0].length; x += 2) {
                Point current = new Point(x, y);
                Point[] adjacents = PointGenerator.adjacents(current);
                Point[] paths = Arrays.stream(adjacents)
                    .filter(p -> tiles()[p.y][p.x])
                    .toArray(Point[]::new);
                if (paths.length == 1) {
                    Point[] candidates = Arrays.stream(adjacents)
                        .filter(p -> !tiles()[p.y][p.x])
                        .map(p -> new Point(2 * p.x - current.x, 2 * p.y - current.y))
                        .toArray(Point[]::new);
                    Point p = farthestPath(current, candidates);
                    System.out.println(current);
                    tiles_[(p.y + current.y) / 2][(p.x + current.x) / 2] = true;
                }
            }
        }
    }

    private Point farthestPath(Point start, Point[] paths) {
        ArrayDeque<PointNode> queue = new ArrayDeque<>();
        int[][] pathLengths = new int[Constants.Game.ROWS][Constants.Game.COLUMNS];
        queue.add(new PointNode(null, start, new PointNode.FScore(1, 0)));

        while (!queue.isEmpty()) {
            PointNode currentNode = queue.pollFirst();
            Point current = currentNode.value();
            if (pathLengths[current.y][current.x] != 0) {
                continue;
            }
            pathLengths[current.y][current.x] = currentNode.score().value();
            Arrays.stream(PointGenerator.adjacents(current))
                .filter(this::hasPathAt)
                .map(p ->
                    new PointNode(
                        null,
                        p,
                        new PointNode.FScore(
                            currentNode.score().gScore() + 1,
                            0)))
                .forEach(queue::add);
        }
        return Arrays.stream(paths)
            .sorted((p1, p2) -> pathLengths[p2.y][p2.x] - pathLengths[p1.y][p1.x])
            .findFirst().orElse(null);
    }

    private void makePath(Point p1, Point p2) {
        Point wall = new Point((p1.x + p2.x) / 2, (p1.y + p2.y) / 2);
        tiles_[wall.y][wall.x] = true;
        tiles_[p2.y][p2.x] = true;
    }

    private Point[] nextPaths(Point point) {
        Point[] neighbors = new Point[]{
            new Point(point.x - 2, point.y),
            new Point(point.x + 2, point.y),
            new Point(point.x, point.y - 2),
            new Point(point.x, point.y + 2)
        };
        return Arrays.stream(neighbors)
            .filter(p -> contains(p) && !tiles_[p.y][p.x])
            .toArray(Point[]::new);
    }
}

File: ts/maze.ts
namespace Maze {
    export class Point {
        constructor(public row: number, public column: number) {}

        adjacents(): Point[] {
            return [
                new Point(this.row - 1, this.column),
                new Point(this.row + 1, this.column),
                new Point(this.row, this.column - 1),
                new Point(this.row, this.column + 1)
            ];
        }

        copy(): Point {
            return new Point(this.row, this.column);
        }

        equals(other: Point): boolean {
            return this.row == other.row && this.column == other.column;
        }
    }

    export class Maze {
        constructor(public rows: number, public columns: number) {}

        contains(p: Point): boolean {
            return p.row >= 0 && p.row < this.rows &&
                p.column >= 0 && p.column < this.columns;
        }

        pathAt(p: Point): boolean {
            return this.contains(p) && this.tiles[p.row][p.column];
        }

        generate(): void {
            this.resetTiles();
            this.generatePaths();
            this.patchDeadEnds();
        }

        private tiles: boolean[][];

        private resetTiles(): void {
            this.tiles = []
            for (let row: number = 0; row < this.rows; ++row) {
                this.tiles[row] = [];
                for (let col: number = 0; col < this.columns; ++col) {
                    this.tiles[row][col] = false;
                }
            }
        }

        private unbuiltPaths(p: Point): Point[] {
            let paths: Point[] = [
                new Point(p.row - 2, p.column),
                new Point(p.row + 2, p.column),
                new Point(p.row, p.column - 2),
                new Point(p.row, p.column + 2)
            ];
            return paths.filter(point =>
                                this.contains(point) && !this.pathAt(point));
        }

        private makePath(start: Point, end: Point): void {
            let wall: Point = new Point(
                Math.floor((start.row + end.row) / 2),
                Math.floor((start.column + end.column) / 2)
            );

            this.tiles[start.row][start.column] = true;
            this.tiles[wall.row][wall.column] = true;
            this.tiles[end.row][end.column] = true;
        }

        private generatePaths(): void {
            let start: Point = new Point(0, 0);
            let stack: Point[] = [start];

            this.tiles[start.row][start.column] = true;

            while (stack.length > 0) {
                let current: Point = stack[stack.length - 1];
                let nexts: Point[] = this.unbuiltPaths(current);

                if (nexts.length > 0) {
                    let i: number = Math.floor(Math.random() * nexts.length);
                    let next: Point = nexts[i];
                    this.makePath(current, next);
                    stack.push(next);
                } else {
                    stack.pop();
                }
            }
        }

        private pathDistance(start: Point, end: Point): number {
            if (!this.pathAt(start) || !(this.pathAt(end))) {
                return -1;
            }

            let queue: [Point, number][] = [[start, 0]];
            let visited: Point[] = [];

            while (queue.length > 0) {
                let point: Point, distance: number;
                [point, distance] = queue.shift();

                if (point.equals(end)) {
                    return distance;
                }

                for (let neighbor of point.adjacents().filter(p => this.pathAt(p))) {
                    if (!visited.some(p => neighbor.equals(p))) {
                        queue.push([neighbor, distance + 1]);
                    }
                }

                visited.push(point);
            }
        }

        private farthestPoint(start: Point, points: Point[]): Point {
            let distances: [Point, number][] = points.map(
                p => [p, this.pathDistance(start, p)] as [Point, number]);
            return distances.sort((lhs, rhs) => rhs[1] - lhs[1])[0][0];
        }

        private deadEndAt(point: Point): boolean {
            let adjacentPaths: Point[] = point.adjacents()
                .filter(p => this.pathAt(p));
            return adjacentPaths.length == 1;
        }

        private patchDeadEnds(): void {
            for (let row: number = 0; row < this.rows; row += 2) {
                for (let col: number = 0; col < this.columns; col += 2) {
                    let point: Point = new Point(row, col);

                    if (this.deadEndAt(point)) {
                        let candidates: Point[] = [
                            new Point(point.row - 2, point.column),
                            new Point(point.row + 2, point.column),
                            new Point(point.row, point.column - 2),
                            new Point(point.row, point.column + 2)
                        ].filter(p => this.pathAt(p));

                        this.makePath(point, this.farthestPoint(point, candidates));
                    }
                }
            }
        }
    }
}
